chalice/deploy/newdeployer.py

```diff
@@ -0,0 +1,451 @@
+replaced those ``models.DeployPhase.BUILD`` values with whatever the "built"
+value is (e.g the filename of the zipped deployment package).
+
+For example, we know when we create a lambda function that we need to create
+a deployment package, but we don't know the name nor contents of the
+deployment package until the ``LambdaDeploymentPackager`` runs.  Therefore,
+the Resource Builder stage can record the fact that it knows that a
+``models.DeploymentPackage`` is needed, but use ``models.DeployPhase.BUILD``
+for the value of the filename.  The enum values aren't strictly necessary,
+they just add clarity about when this value is expected to be filled in.
+These could also just be set to ``None`` and be of type ``Optional[T]``.
+
+
+Execution Plan Stage
+====================
+
+This stage takes the ordered list of resources and figures out what AWS API
+calls we have to make.  For example, if a resource doesn't exist at all, we'll
+need to make a ``create_*`` call.  If the resource exists, we may need to make
+a series of ``update_*`` calls.  If the resource exists and is already up to
+date, we might not need to make any calls at all.  The output of this stage is
+a list of ``APICall`` objects.  This stage doesn't actually make the mutating
+API calls, it only figures out what calls we should make.  This stage will
+typically only make ``describe/list`` AWS calls.
+
+
+The Executor
+============
+
+This takes the list of ``APICall`` objects from the previous stage and finally
+executes them.  It also manages taking the output of API calls and storing them
+in variables so they can be referenced in subsequent ``APICall`` objects (see
+the ``Variable`` class to see how this is used).  For example, if a lambda
+function needs the ``role_arn`` that's the result of a previous ``create_role``
+API call, a ``Variable`` object is used to forward this information.
+
+The executor also records these variables with their associated resources so a
+``deployed.json`` file can be written to disk afterwards.  An ``APICall``
+takes an optional resource object when it's created whose ``resource_name``
+is used as the key in the ``deployed.json`` dictionary.
+
+
+"""
+import os
+
+from typing import List, Set, Dict, Any, Optional, Union  # noqa
+from botocore.session import Session  # noqa
+
+from chalice.utils import OSUtils, UI
+from chalice.deploy import models
+from chalice.config import Config  # noqa
+from chalice import app  # noqa
+from chalice.deploy.packager import LambdaDeploymentPackager
+from chalice.deploy.packager import PipRunner, SubprocessPip
+from chalice.deploy.packager import DependencyBuilder as PipDependencyBuilder
+from chalice.policy import AppPolicyGenerator
+from chalice.constants import LAMBDA_TRUST_POLICY
+from chalice.constants import DEFAULT_LAMBDA_TIMEOUT
+from chalice.constants import DEFAULT_LAMBDA_MEMORY_SIZE
+from chalice.awsclient import TypedAWSClient, ResourceDoesNotExistError
+
+
+def create_default_deployer(session):
```

Comment:
Capturing offline discussion about changing this to use a builder pattern instead.

Reply:
Yep, though I'm going to hold off on that until we need it for deletes.  For now a static factory function is sufficient and simpler to follow.